woocommerce/woocommerce-gutenberg-products-block
# Add Formatter Classes to the Store API for Extensions to Consume

This is to help implement subscriptions support (https://github.com/woocommerce/woocommerce-subscriptions/pull/3904).

When extending the REST API, extensions may need access to certain formatting functions to avoid re-implementing them in the extension itself. This applies mainly to monetary and currency values where our Store API uses specific formats.

This PR attempts to solve this by exposing formatter classes via the `ExtendRestAPI` class. I've included currency, money, and HTML formatters in this PR.

Example usage:

```php
$this->extend->formatters->money->format( $value )
```

Formatters are registered via Bootstrap and injected into the `ExtendRestAPI` class.

Each individual formatter has public methods to `format`, `set_option`, and `get_option`. This ensures all formatters are compatible with each other, and in the absence of a specific formatter there are no fatal errors from calling undefined methods.

### How to test the changes in this Pull Request:

If you do a GET request on the cart endpoint, you should see correct results. This endpoint has implemented the new formatters via `prepare_money_response`.

I have included tests to check formatter functionality, and registration.

### Changelog

> Added formatting classes to the Store API for extensions to consume.

cc'ing @nerrad to ensure he is happy with this pattern of usage.
